intake/source/cache.py

```diff
@@ -0,0 +1,228 @@
+
+        Parameters:
+        ----------
+        urlpath: str, location of data
+            May be a local path, or remote path if including a protocol specifier
+            such as ``'s3://'``. May include glob wildcards.
+
+        Returns
+        -------
+        List of local cache_paths to be opened instead of the remote file(s). If
+        caching is disable, the urlpath is returned.
+        """
+        if os.getenv('INTAKE_DISABLE_CACHING') == '1':
+            return [urlpath]
+
+        from dask.bytes import open_files
+
+        self._ensure_cache_dir()
+        subdir = self._hash(urlpath)
+        cache_paths = []
+        files_in = open_files(urlpath, 'rb')
+        files_out = open_files([self._path(f.path, subdir) for f in files_in], 'wb')
+        for file_in, file_out in zip(files_in, files_out):
+            cache_path = file_out.path
+            cache_paths.append(cache_path)
+
+            if not os.path.isfile(cache_path):
+                logger.info("Caching file {} from urlpath {}".format(file_in.path, urlpath))
+                self._log_metadata(urlpath, file_in.path, cache_path)
+
+                with file_in as f1:
+                    with file_out as f2:
+                        data = True
+                        while data:
+                            #TODO: print out progress
+                            data = f1.read(self.blocksize)
+                            f2.write(data)
+        return cache_paths
+
+    def get_metadata(self, urlpath):
+        """
+        Parameters:
+        ----------
+        urlpath: str, location of data
+            May be a local path, or remote path if including a protocol specifier
+            such as ``'s3://'``. May include glob wildcards.
+
+        Returns
+        -------
+        Metadata (dict) about a given urlpath.
+        """
+        return self._metadata[urlpath]
+    
+    def clear_cache(self, urlpath):
+        """
+        Clears cache and metadata for a given urlpath.
+
+        Parameters:
+        ----------
+        urlpath: str, location of data
+            May be a local path, or remote path if including a protocol specifier
+            such as ``'s3://'``. May include glob wildcards.
+        """
+        cache_entries = self._metadata.pop(urlpath)
+        for cache_entry in cache_entries:
+            try:
+                os.remove(cache_entry['cache_path'])
+            except FileNotFoundError:
+                pass
+        os.rmdir(os.path.dirname(cache_entry['cache_path']))
+    
+    def clear_all(self):
+        """
+        Clears all cache and metadata.
+        """
+        for urlpath in self._metadata.keys():
+            self.clear_cache(urlpath)
+        
+        # Safely clean up anything else.
+        try:
+            for subdir in os.listdir(self._cache_dir):
+                shutil.rmtree(os.path.join(self._cache_dir, subdir))
+        except FileNotFoundError:
+            pass
+
+
+class CacheMetadata(object):
```

Comment:
Derive from dict or collections.MutableMapping?
May want to fill out API. In fact, would be surprised if an interface to on-disc JSON dict didn't already exist, maybe with nice thread-safe behaviour for free - worth a search.

Reply:
Yes, I realize this is starting get out of control. I can at least derive from dict. I had thought about using [shelve](https://docs.python.org/2/library/shelve.html) to persist, however its not very human readable as a pickle without loading it.